### PR TITLE
drivers: usb_dc_nrfx: Fix improper memory sizes

### PR DESCRIPTION
**Pull request by MarekPieta**
Change switches to using static memory for out endpoints and mem_slab for fifo elements. This fixes problems related to improper heap memory size.

### CODE REVIEW

File: drivers/usb/device/usb_dc_nrfx.c [hunk 1]
```diff
@@ -168,22 +168,21 @@ struct usbd_event {
...
  * on the number of endpoints configured.
  */
-#define FIFO_ELEM_MIN_SZ        sizeof(struct usbd_event)
-#define FIFO_ELEM_MAX_SZ        sizeof(struct usbd_event)
+#define FIFO_ELEM_SZ            sizeof(struct usbd_event)
 #define FIFO_ELEM_ALIGN         sizeof(unsigned int)
 
 #if CONFIG_USB_NRFX_EVT_QUEUE_SIZE < 4
 #error Invalid USBD event queue size (CONFIG_USB_NRFX_EVT_QUEUE_SIZE).
 #endif
```
Comment (emob-nordic) [inline drivers/usb/device/usb_dc_nrfx.c:182]: I know its not part of this PR but please remove this if You can. Its not needed, Kconfig will throw an error when invalid value is assigned. This option has `range 4 64`.

**Comment by jfischer-no** [comment]
An other regression introduced by #28611

File: drivers/usb/device/usb_dc_nrfx.c [hunk 2]
```diff
@@ -212,34 +207,20 @@ K_HEAP_DEFINE(fifo_elem_pool,
...
-#endif
+/**
+ * @brief Output endpoint buffers
+ *	Used as buffers for the endpoints' data transfer
+ *	Max buffers size possible: 2048 Bytes (16 EP * 64B + 1 ISO * 1024B)
+ */
+static uint8_t ep_out_bufs[CFG_EPOUT_CNT][EP_BUF_MAX_SZ]
+	       __aligned(sizeof(unsigned int));
+static uint8_t ep_isoout_buf[ISO_EP_BUF_MAX_SZ]
+	       __aligned(sizeof(unsigned int));
```
Comment (emob-nordic) [inline drivers/usb/device/usb_dc_nrfx.c:217]: I would prefer `CFG_EPISOOUT_CNT` as well. Update allocator as well.

**Review by eobalski — CHANGES_REQUESTED** [review]
Only 2 minor things.

File: drivers/usb/device/usb_dc_nrfx.c (hunk 2 above)
Comment (MarekPieta) [reply drivers/usb/device/usb_dc_nrfx.c:217]: done

File: drivers/usb/device/usb_dc_nrfx.c (hunk 1 above)
Comment (MarekPieta) [reply drivers/usb/device/usb_dc_nrfx.c:182]: done

**Review by eobalski — APPROVED** [review]

File: drivers/usb/device/usb_dc_nrfx.c [hunk 3]
```diff
@@ -453,10 +444,7 @@ static inline struct usbd_event *usbd_evt_alloc(void)
 		 */
 		usbd_evt_flush();
 
-		block.data = k_heap_alloc(&fifo_elem_pool,
-					   sizeof(struct usbd_event),
-					   K_NO_WAIT);
-		if (block.data == NULL) {
+		if (k_mem_slab_alloc(&fifo_elem_slab, (void **)&block.data, K_NO_WAIT)) {
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_nrfx.c:447]: Please reshape.
```diff c
diff --git a/drivers/usb/device/usb_dc_nrfx.c b/drivers/usb/device/usb_dc_nrfx.c
index 7ce5e66b60..2c842f037a 100644
--- a/drivers/usb/device/usb_dc_nrfx.c
+++ b/drivers/usb/device/usb_dc_nrfx.c
@@ -444,7 +444,8 @@ static inline struct usbd_event *usbd_evt_alloc(void)
                 */
                usbd_evt_flush();
 
-               if (k_mem_slab_alloc(&fifo_elem_slab, (void **)&block.data, K_NO_WAIT)) {
+               if (k_mem_slab_alloc(&fifo_elem_slab,
+                                    (void **)&block.data, K_NO_WAIT)) {
                        LOG_ERR("USBD event memory corrupted");
                        __ASSERT_NO_MSG(0);
                        return NULL;
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Thanks, just few comments about style.

File: drivers/usb/device/usb_dc_nrfx.c [hunk 4]
```diff
@@ -212,34 +207,20 @@ K_HEAP_DEFINE(fifo_elem_pool,
...
-#else
-#define EP_BUF_POOL_BLOCK_MAX_SZ EP_BUF_MAX_SZ
-#endif
+/**
+ * @brief Output endpoint buffers
+ *	Used as buffers for the endpoints' data transfer
+ *	Max buffers size possible: 2048 Bytes (16 EP * 64B + 1 ISO * 1024B)
+ */
+static uint8_t ep_out_bufs[CFG_EPOUT_CNT][EP_BUF_MAX_SZ]
+	       __aligned(sizeof(unsigned int));
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply drivers/usb/device/usb_dc_nrfx.c:211]: done

File: drivers/usb/device/usb_dc_nrfx.c [hunk 5]
```diff
@@ -212,34 +207,20 @@ K_HEAP_DEFINE(fifo_elem_pool,
...
-/** Maximum endpoint buffer size (maximum block size) */
-#if (CFG_EP_ISOIN_CNT || CFG_EP_ISOOUT_CNT)
-#define EP_BUF_POOL_BLOCK_MAX_SZ ISO_EP_BUF_MAX_SZ
-#else
-#define EP_BUF_POOL_BLOCK_MAX_SZ EP_BUF_MAX_SZ
-#endif
+/**
+ * @brief Output endpoint buffers
+ *	Used as buffers for the endpoints' data transfer
+ *	Max buffers size possible: 2048 Bytes (16 EP * 64B + 1 ISO * 1024B)
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply drivers/usb/device/usb_dc_nrfx.c:208]: Probably that's a leftover from previous implementation that allocated buffers also for IN endpoints. Thanks for clarification. Fixed.

File: drivers/usb/device/usb_dc_nrfx.c (hunk 3 above)
Comment (MarekPieta) [reply drivers/usb/device/usb_dc_nrfx.c:447]: done

**Review by jfischer-no — APPROVED** [review]
